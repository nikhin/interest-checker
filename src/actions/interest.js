import { FETCH_INTEREST } from './constants';

export const fetchInterest = () => {
  return dispatch => {
    return fetch('http://localhost:3001/interest')
      .then(response => response.json())
      .then(json => dispatch({ type: FETCH_INTEREST, interest: json }));
  }
};