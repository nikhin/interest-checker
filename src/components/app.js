import React from 'react';
import BankBalance from './bank-balance';
import Interest from './interest';

const App = () => {
  return (
    <div>
      <h2>Interest Calculator</h2>
      <hr />
      <BankBalance />
      <hr />
      <Interest />
      
    </div>
  )
}

export default App;