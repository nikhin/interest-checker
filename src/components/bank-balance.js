import React, { Component } from 'react';
import { connect } from 'react-redux';
import { deposit, withdraw } from '../actions/balance';

export class BankBalance extends Component {
  constructor() {
    super();

    this.state = { balance: 0 };
  }

  updateBalance = event => this.setState({ balance: parseInt(event.target.value) })

  deposit = () => this.props.deposit(this.state.balance)

  withdraw = () => this.props.withdraw(this.state.balance)

  render() {
    return (
      <div>
        <h3 className='balance'>Bank balance: {this.props.balance}</h3>
        <br />
        <input className='input-balance' onChange={this.updateBalance} />
        <button className='btn-deposit' onClick={this.deposit}>Deposit</button>
        <button className='btn-withdraw' onClick={this.withdraw}>Withdraw</button>
      </div>
    )
  }
}

export default connect(state => { return { balance: state.balance } }, { deposit, withdraw })(BankBalance);