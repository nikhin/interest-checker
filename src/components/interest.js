import React, { Component } from 'react';
import { connect } from 'react-redux';
import { fetchInterest } from '../actions/interest';
import _ from 'lodash'

export class Interest extends Component {
  componentDidMount() {
    this.props.fetchInterest();
  }

  computeInterest() {
    const { interest } = this.props;

    if (Object.keys(interest).length === 0) return '';

    return _.round(this.props.balance * parseFloat(interest.finance.interestValue),2);
  }

  render() {
    return (
      <h3>Total Interest: {this.computeInterest()}</h3>
    )
  }
}

export default connect(state => state, { fetchInterest })(Interest);