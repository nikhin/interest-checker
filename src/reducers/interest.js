import { FETCH_INTEREST } from '../actions/constants';

const interest = (state = {}, action) => {
  switch(action.type) {
    case FETCH_INTEREST:
      return action.interest;
    default:
      return state;
  }
};

export default interest;