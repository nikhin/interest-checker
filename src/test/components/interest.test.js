import React from 'react';
import { mount, shallow } from 'enzyme';
import { Interest } from '../../components/interest';

describe('Interest', () => {
  let props = { balance: 10, interest: {} };
  let interest = shallow(<Interest {...props} />);

  it('renders properly', () => {
    expect(interest).toMatchSnapshot();
  });

  describe('when mounted', () => {
    const mockFetchinterest = jest.fn();

    beforeEach(() => {
      props.fetchInterest = mockFetchinterest;
      interest = mount(<Interest {...props} />);
    });

    it('dispatches the `fetchInterest()` method it receives from props', () => {
      expect(mockFetchinterest).toHaveBeenCalled();
    });
  });

  describe('when there are valid interest props', () => {
    beforeEach(() => {
      props = { balance: 10, interest: {"finance": {"interestValue": "0.09","bank": "rbi","location": "mississauga"}} };
      interest = shallow(<Interest {...props} />);
    });

    it('displays the correct interest value', () => {
      expect(interest.find('h3').text()).toEqual('Total Interest: 0.9');
    });
  });
});