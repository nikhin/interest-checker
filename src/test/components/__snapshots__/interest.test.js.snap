// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Interest renders properly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Interest
    balance={10}
    interest={Object {}}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        "Total Interest: ",
        "",
      ],
    },
    "ref": null,
    "rendered": Array [
      "Total Interest: ",
      "",
    ],
    "type": "h3",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          "Total Interest: ",
          "",
        ],
      },
      "ref": null,
      "rendered": Array [
        "Total Interest: ",
        "",
      ],
      "type": "h3",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
